src/util/taproot.rs

```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct NodeInfo {
+    /// Merkle Hash for this node
+    hash: sha256::Hash,
+    /// information about leaves inside this node
+    leaves: Vec<LeafInfo>,
+}
+
+impl NodeInfo {
+    // Create a new NodeInfo with omitted/hidden info
+    fn new_hidden(hash: sha256::Hash) -> Self {
+        Self {
+            hash: hash,
+            leaves: vec![],
+        }
+    }
+
+    // Create a new leaf with NodeInfo
+    fn new_leaf_with_ver(script: Script, ver: LeafVersion) -> Self {
+        let leaf = LeafInfo::new(script, ver);
+        Self {
+            hash: leaf.hash(),
+            leaves: vec![leaf],
+        }
+    }
+
+    // Combine two NodeInfo's to create a new parent
+    fn combine(a: Self, b: Self) -> Result<Self, TaprootBuilderError> {
+        let mut all_leaves = Vec::with_capacity(a.leaves.len() + b.leaves.len());
+        for mut a_leaf in a.leaves {
+            a_leaf.merkle_branch.push(b.hash)?; // add hashing partner
+            all_leaves.push(a_leaf);
+        }
+        for mut b_leaf in b.leaves {
+            b_leaf.merkle_branch.push(a.hash)?; // add hashing partner
+            all_leaves.push(b_leaf);
+        }
+        let mut eng = TapBranchHash::engine();
+        if a.hash < b.hash {
+            eng.input(&a.hash);
+            eng.input(&b.hash);
+        } else {
+            eng.input(&b.hash);
+            eng.input(&a.hash);
+        };
+        Ok(Self {
+            hash: sha256::Hash::from_engine(eng),
+            leaves: all_leaves,
+        })
+    }
+}
+
+// Internally used structure to store information about taproot leaf node
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct LeafInfo {
+    /// The underlying script
+    script: Script,
+    /// The leaf version
+    ver: LeafVersion,
+    /// The merkle proof(hashing partners) to get this node
+    merkle_branch: TaprootMerkleBranch,
+}
+
+impl LeafInfo {
+    // Create an instance of Self from Script with default version and no merkle branch
+    fn new(script: Script, ver: LeafVersion) -> Self {
+        Self {
+            script: script,
+            ver: ver,
+            merkle_branch: TaprootMerkleBranch(vec![]),
+        }
+    }
+
+    // Compute a leaf hash for the given leaf
+    fn hash(&self) -> sha256::Hash {
+        compute_leaf_hash(&self.script, self.ver)
+    }
+}
+
+/// function to compute leaf hash from components
+pub fn compute_leaf_hash(script: &Script, ver: LeafVersion) -> sha256::Hash {
+    let mut eng = TapLeafHash::engine();
+    ver.as_u8()
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    script
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    sha256::Hash::from_engine(eng)
+}
+
+/// helper function to obtain tap tweak
+pub fn compute_tap_tweak(
+    merkle_root: Option<sha256::Hash>,
+    internal_key: schnorr::PublicKey,
+) -> sha256::Hash {
```

Comment:
Why we do not return `TapTweakHash` here?

Reply:
I should, I will change the methods to operate on correct hashes, taptweakhash, tapbranchhash etc.